File: app/src/main/java/org/schabi/newpipe/settings/PeertubeInstanceListFragment.java
package org.schabi.newpipe.settings;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.text.InputType;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.RadioButton;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.content.res.AppCompatResources;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.grack.nanojson.JsonStringWriter;
import com.grack.nanojson.JsonWriter;

import org.schabi.newpipe.R;
import org.schabi.newpipe.extractor.services.peertube.PeertubeInstance;
import org.schabi.newpipe.util.Constants;
import org.schabi.newpipe.util.PeertubeHelper;
import org.schabi.newpipe.util.ThemeHelper;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

public class PeertubeInstanceListFragment extends Fragment {
    private static final int MENU_ITEM_RESTORE_ID = 123456;

    private List<PeertubeInstance> instanceList = new ArrayList<>();
    private PeertubeInstance selectedInstance;
    private String savedInstanceListKey;
    private InstanceListAdapter instanceListAdapter;

    private ProgressBar progressBar;
    private SharedPreferences sharedPreferences;

    private CompositeDisposable disposables = new CompositeDisposable();

    /*//////////////////////////////////////////////////////////////////////////
    // Lifecycle
    //////////////////////////////////////////////////////////////////////////*/

    @Override
    public void onCreate(@Nullable final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(requireContext());
        savedInstanceListKey = getString(R.string.peertube_instance_list_key);
        selectedInstance = PeertubeHelper.getCurrentInstance();
        updateInstanceList();

        setHasOptionsMenu(true);
    }

    @Override
    public View onCreateView(@NonNull final LayoutInflater inflater, final ViewGroup container,
                             final Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_instance_list, container, false);
    }

    @Override
    public void onViewCreated(@NonNull final View rootView,
                              @Nullable final Bundle savedInstanceState) {
        super.onViewCreated(rootView, savedInstanceState);

        initViews(rootView);
    }

    private void initViews(@NonNull final View rootView) {
        TextView instanceHelpTV = rootView.findViewById(R.id.instanceHelpTV);
        instanceHelpTV.setText(getString(R.string.peertube_instance_url_help,
                getString(R.string.peertube_instance_list_url)));

        initButton(rootView);

        RecyclerView listInstances = rootView.findViewById(R.id.instances);
        listInstances.setLayoutManager(new LinearLayoutManager(requireContext()));

        ItemTouchHelper itemTouchHelper = new ItemTouchHelper(getItemTouchCallback());
        itemTouchHelper.attachToRecyclerView(listInstances);

        instanceListAdapter = new InstanceListAdapter(requireContext(), itemTouchHelper);
        listInstances.setAdapter(instanceListAdapter);

        progressBar = rootView.findViewById(R.id.loading_progress_bar);
    }

    @Override
    public void onResume() {
        super.onResume();
        updateTitle();
    }

    @Override
    public void onPause() {
        super.onPause();
        saveChanges();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (disposables != null) {
            disposables.clear();
        }
        disposables = null;
    }

    /*//////////////////////////////////////////////////////////////////////////
    // Menu
    //////////////////////////////////////////////////////////////////////////*/

    @Override
    public void onCreateOptionsMenu(final Menu menu, final MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);

        final MenuItem restoreItem = menu
                .add(Menu.NONE, MENU_ITEM_RESTORE_ID, Menu.NONE, R.string.restore_defaults);
        restoreItem.setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS);

        final int restoreIcon = ThemeHelper
                .resolveResourceIdFromAttr(requireContext(), R.attr.ic_restore_defaults);
        restoreItem.setIcon(AppCompatResources.getDrawable(requireContext(), restoreIcon));
    }

    @Override
    public boolean onOptionsItemSelected(final MenuItem item) {
        if (item.getItemId() == MENU_ITEM_RESTORE_ID) {
            restoreDefaults();
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    /*//////////////////////////////////////////////////////////////////////////
    // Utils
    //////////////////////////////////////////////////////////////////////////*/

    private void updateInstanceList() {
        instanceList.clear();
        instanceList.addAll(PeertubeHelper.getInstanceList(requireContext()));
    }

    private void selectInstance(final PeertubeInstance instance) {
        selectedInstance = PeertubeHelper.selectInstance(instance, requireContext());
        sharedPreferences.edit().putBoolean(Constants.KEY_MAIN_PAGE_CHANGE, true).apply();
    }

    private void updateTitle() {
        if (getActivity() instanceof AppCompatActivity) {
            ActionBar actionBar = ((AppCompatActivity) getActivity()).getSupportActionBar();
            if (actionBar != null) {
                actionBar.setTitle(R.string.peertube_instance_url_title);
            }
        }
    }

    private void saveChanges() {
        JsonStringWriter jsonWriter = JsonWriter.string().object().array("instances");
        for (PeertubeInstance instance : instanceList) {
            jsonWriter.object();
            jsonWriter.value("name", instance.getName());
            jsonWriter.value("url", instance.getUrl());
            jsonWriter.end();
        }
        String jsonToSave = jsonWriter.end().end().done();
        sharedPreferences.edit().putString(savedInstanceListKey, jsonToSave).apply();
    }

    private void restoreDefaults() {
        new AlertDialog.Builder(requireContext(), ThemeHelper.getDialogTheme(requireContext()))
                .setTitle(R.string.restore_defaults)
                .setMessage(R.string.restore_defaults_confirmation)
                .setNegativeButton(R.string.cancel, null)
                .setPositiveButton(R.string.yes, (dialog, which) -> {
                    sharedPreferences.edit().remove(savedInstanceListKey).apply();
                    selectInstance(PeertubeInstance.defaultInstance);
                    updateInstanceList();
                    instanceListAdapter.notifyDataSetChanged();
                })
                .show();
    }

    private void initButton(final View rootView) {
        final FloatingActionButton fab = rootView.findViewById(R.id.addInstanceButton);
        fab.setOnClickListener(v -> {
            showAddItemDialog(requireContext());
        });
    }

    private void showAddItemDialog(final Context c) {
        final EditText urlET = new EditText(c);
        urlET.setInputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_URI);
        urlET.setHint(R.string.peertube_instance_add_help);
        AlertDialog dialog = new AlertDialog.Builder(c)
                .setTitle(R.string.peertube_instance_add_title)
                .setIcon(R.drawable.place_holder_peertube)
                .setNegativeButton(R.string.cancel, null)
                .setPositiveButton(R.string.finish, (dialog1, which) -> {
                    String url = urlET.getText().toString();
                    addInstance(url);
                })
                .create();
        dialog.setView(urlET, 50, 0, 50, 0);
        dialog.show();
    }

    private void addInstance(final String url) {
        String cleanUrl = cleanUrl(url);
        if (cleanUrl == null) {
            return;
        }
        progressBar.setVisibility(View.VISIBLE);
        Disposable disposable = Single.fromCallable(() -> {
            PeertubeInstance instance = new PeertubeInstance(cleanUrl);
            instance.fetchInstanceMetaData();
            return instance;
        }).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
                .subscribe((instance) -> {
                    progressBar.setVisibility(View.GONE);
                    add(instance);
                }, e -> {
                    progressBar.setVisibility(View.GONE);
                    Toast.makeText(getActivity(), R.string.peertube_instance_add_fail,
                            Toast.LENGTH_SHORT).show();
                });
        disposables.add(disposable);
    }

    @Nullable
    private String cleanUrl(final String url) {
        String cleanUrl = url.trim();
        // if protocol not present, add https
        if (!cleanUrl.startsWith("http")) {
            cleanUrl = "https://" + cleanUrl;
        }
        // remove trailing slash
        cleanUrl = cleanUrl.replaceAll("/$", "");
        // only allow https
        if (!cleanUrl.startsWith("https://")) {
            Toast.makeText(getActivity(), R.string.peertube_instance_add_https_only,
                    Toast.LENGTH_SHORT).show();
            return null;
        }
        // only allow if not already exists
        for (PeertubeInstance instance : instanceList) {
            if (instance.getUrl().equals(cleanUrl)) {
                Toast.makeText(getActivity(), R.string.peertube_instance_add_exists,
                        Toast.LENGTH_SHORT).show();
                return null;
            }
        }
        return cleanUrl;
    }

    private void add(final PeertubeInstance instance) {
        instanceList.add(instance);
        instanceListAdapter.notifyDataSetChanged();
    }

    private ItemTouchHelper.SimpleCallback getItemTouchCallback() {
        return new ItemTouchHelper.SimpleCallback(ItemTouchHelper.UP | ItemTouchHelper.DOWN,
                ItemTouchHelper.START | ItemTouchHelper.END) {
            @Override
            public int interpolateOutOfBoundsScroll(final RecyclerView recyclerView,
                                                    final int viewSize,
                                                    final int viewSizeOutOfBounds,
                                                    final int totalSize,
                                                    final long msSinceStartScroll) {
                final int standardSpeed = super.interpolateOutOfBoundsScroll(recyclerView, viewSize,
                        viewSizeOutOfBounds, totalSize, msSinceStartScroll);
                final int minimumAbsVelocity = Math.max(12,
                        Math.abs(standardSpeed));
                return minimumAbsVelocity * (int) Math.signum(viewSizeOutOfBounds);
            }

            @Override
            public boolean onMove(final RecyclerView recyclerView,
                                  final RecyclerView.ViewHolder source,
                                  final RecyclerView.ViewHolder target) {
                if (source.getItemViewType() != target.getItemViewType()
                        || instanceListAdapter == null) {
                    return false;
                }

                final int sourceIndex = source.getAdapterPosition();
                final int targetIndex = target.getAdapterPosition();
                instanceListAdapter.swapItems(sourceIndex, targetIndex);
                return true;
            }

            @Override
            public boolean isLongPressDragEnabled() {
                return false;
            }

            @Override
            public boolean isItemViewSwipeEnabled() {
                return true;
            }

            @Override
            public void onSwiped(final RecyclerView.ViewHolder viewHolder, final int swipeDir) {
                int position = viewHolder.getAdapterPosition();
                // do not allow swiping the selected instance
                if (instanceList.get(position).getUrl().equals(selectedInstance.getUrl())) {
                    instanceListAdapter.notifyItemChanged(position);
                    return;
                }
                instanceList.remove(position);
                instanceListAdapter.notifyItemRemoved(position);

                if (instanceList.isEmpty()) {
                    instanceList.add(selectedInstance);
                    instanceListAdapter.notifyItemInserted(0);
                }
            }
        };
    }

    /*//////////////////////////////////////////////////////////////////////////
    // List Handling
    //////////////////////////////////////////////////////////////////////////*/

    private class InstanceListAdapter
            extends RecyclerView.Adapter<InstanceListAdapter.TabViewHolder> {
        private final LayoutInflater inflater;
        private ItemTouchHelper itemTouchHelper;
        private RadioButton lastChecked;

        InstanceListAdapter(final Context context, final ItemTouchHelper itemTouchHelper) {
            this.itemTouchHelper = itemTouchHelper;
            this.inflater = LayoutInflater.from(context);
        }

        public void swapItems(final int fromPosition, final int toPosition) {
            Collections.swap(instanceList, fromPosition, toPosition);
            notifyItemMoved(fromPosition, toPosition);
        }

        @NonNull
        @Override
        public InstanceListAdapter.TabViewHolder onCreateViewHolder(@NonNull final ViewGroup parent,
                                                                    final int viewType) {
            View view = inflater.inflate(R.layout.item_instance, parent, false);
            return new InstanceListAdapter.TabViewHolder(view);
        }

        @Override
        public void onBindViewHolder(@NonNull final InstanceListAdapter.TabViewHolder holder,
                                     final int position) {
            holder.bind(position, holder);
        }

        @Override
        public int getItemCount() {
            return instanceList.size();
        }

        class TabViewHolder extends RecyclerView.ViewHolder {
            private AppCompatImageView instanceIconView;
            private TextView instanceNameView;
            private TextView instanceUrlView;
            private RadioButton instanceRB;
            private ImageView handle;

            TabViewHolder(final View itemView) {
                super(itemView);

                instanceIconView = itemView.findViewById(R.id.instanceIcon);
                instanceNameView = itemView.findViewById(R.id.instanceName);
                instanceUrlView = itemView.findViewById(R.id.instanceUrl);
                instanceRB = itemView.findViewById(R.id.selectInstanceRB);
                handle = itemView.findViewById(R.id.handle);
            }

            @SuppressLint("ClickableViewAccessibility")
            void bind(final int position, final TabViewHolder holder) {
                handle.setOnTouchListener(getOnTouchListener(holder));

                final PeertubeInstance instance = instanceList.get(position);
                instanceNameView.setText(instance.getName());
                instanceUrlView.setText(instance.getUrl());
                instanceRB.setOnCheckedChangeListener(null);
                if (selectedInstance.getUrl().equals(instance.getUrl())) {
                    if (lastChecked != null && lastChecked != instanceRB) {
                        lastChecked.setChecked(false);
                    }
                    instanceRB.setChecked(true);
                    lastChecked = instanceRB;
                }
                instanceRB.setOnCheckedChangeListener((buttonView, isChecked) -> {
                    if (isChecked) {
                        selectInstance(instance);
                        if (lastChecked != null && lastChecked != instanceRB) {
                            lastChecked.setChecked(false);
                        }
                        lastChecked = instanceRB;
                    }
                });
                instanceIconView.setImageResource(R.drawable.place_holder_peertube);
            }

            @SuppressLint("ClickableViewAccessibility")
            private View.OnTouchListener getOnTouchListener(final RecyclerView.ViewHolder item) {
                return (view, motionEvent) -> {
                    if (motionEvent.getActionMasked() == MotionEvent.ACTION_DOWN) {
                        if (itemTouchHelper != null && getItemCount() > 1) {
                            itemTouchHelper.startDrag(item);
                            return true;
                        }
                    }
                    return false;
                };
            }
        }
    }
}

File: app/src/main/java/org/schabi/newpipe/settings/PeertubeInstanceListFragment.kt
package org.schabi.newpipe.settings

import android.annotation.SuppressLint
import android.content.Context
import android.content.DialogInterface
import android.content.SharedPreferences
import android.os.Bundle
import android.text.InputType
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.MotionEvent
import android.view.View
import android.view.View.OnTouchListener
import android.view.ViewGroup
import android.widget.CompoundButton
import android.widget.RadioButton
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.Fragment
import androidx.preference.PreferenceManager
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.grack.nanojson.JsonStringWriter
import com.grack.nanojson.JsonWriter
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.core.Single
import io.reactivex.rxjava3.disposables.CompositeDisposable
import io.reactivex.rxjava3.disposables.Disposable
import io.reactivex.rxjava3.functions.Consumer
import io.reactivex.rxjava3.schedulers.Schedulers
import org.schabi.newpipe.R
import org.schabi.newpipe.databinding.DialogEditTextBinding
import org.schabi.newpipe.databinding.FragmentInstanceListBinding
import org.schabi.newpipe.databinding.ItemInstanceBinding
import org.schabi.newpipe.extractor.services.peertube.PeertubeInstance
import org.schabi.newpipe.util.PeertubeHelper
import org.schabi.newpipe.util.ThemeHelper
import java.util.Collections
import java.util.concurrent.Callable
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.sign

class PeertubeInstanceListFragment() : Fragment() {
    private var selectedInstance: PeertubeInstance? = null
    private var savedInstanceListKey: String? = null
    private var instanceListAdapter: InstanceListAdapter? = null
    private var binding: FragmentInstanceListBinding? = null
    private var sharedPreferences: SharedPreferences? = null
    private var disposables: CompositeDisposable? = CompositeDisposable()

    /*//////////////////////////////////////////////////////////////////////////
    // Lifecycle
    ////////////////////////////////////////////////////////////////////////// */
    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(requireContext())
        savedInstanceListKey = getString(R.string.peertube_instance_list_key)
        selectedInstance = PeertubeHelper.getCurrentInstance()
        setHasOptionsMenu(true)
    }

    public override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                                     savedInstanceState: Bundle?): View? {
        binding = FragmentInstanceListBinding.inflate(inflater, container, false)
        return binding!!.getRoot()
    }

    public override fun onViewCreated(rootView: View,
                                      savedInstanceState: Bundle?) {
        super.onViewCreated(rootView, savedInstanceState)
        binding!!.instanceHelpTV.setText(getString(R.string.peertube_instance_url_help,
                getString(R.string.peertube_instance_list_url)))
        binding!!.addInstanceButton.setOnClickListener(View.OnClickListener({ v: View? -> showAddItemDialog(requireContext()) }))
        binding!!.instances.setLayoutManager(LinearLayoutManager(requireContext()))
        val itemTouchHelper: ItemTouchHelper = ItemTouchHelper(getItemTouchCallback())
        itemTouchHelper.attachToRecyclerView(binding!!.instances)
        instanceListAdapter = InstanceListAdapter(requireContext(), itemTouchHelper)
        binding!!.instances.setAdapter(instanceListAdapter)
        instanceListAdapter!!.submitList(PeertubeHelper.getInstanceList(requireContext()))
    }

    public override fun onResume() {
        super.onResume()
        ThemeHelper.setTitleToAppCompatActivity(getActivity(),
                getString(R.string.peertube_instance_url_title))
    }

    public override fun onPause() {
        super.onPause()
        saveChanges()
    }

    public override fun onDestroy() {
        super.onDestroy()
        if (disposables != null) {
            disposables!!.clear()
        }
        disposables = null
    }

    public override fun onDestroyView() {
        binding = null
        super.onDestroyView()
    }

    /*//////////////////////////////////////////////////////////////////////////
    // Menu
    ////////////////////////////////////////////////////////////////////////// */
    public override fun onCreateOptionsMenu(menu: Menu,
                                            inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(R.menu.menu_chooser_fragment, menu)
    }

    public override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.getItemId() == R.id.menu_item_restore_default) {
            restoreDefaults()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    /*//////////////////////////////////////////////////////////////////////////
    // Utils
    ////////////////////////////////////////////////////////////////////////// */
    private fun selectInstance(instance: PeertubeInstance?) {
        selectedInstance = PeertubeHelper.selectInstance(instance, requireContext())
        sharedPreferences!!.edit().putBoolean(KEY_MAIN_PAGE_CHANGE, true).apply()
    }

    private fun saveChanges() {
        val jsonWriter: JsonStringWriter = JsonWriter.string().`object`().array("instances")
        for (instance: PeertubeInstance? in instanceListAdapter!!.getCurrentList()) {
            jsonWriter.`object`()
            jsonWriter.value("name", instance!!.getName())
            jsonWriter.value("url", instance.getUrl())
            jsonWriter.end()
        }
        val jsonToSave: String = jsonWriter.end().end().done()
        sharedPreferences!!.edit().putString(savedInstanceListKey, jsonToSave).apply()
    }

    private fun restoreDefaults() {
        val context: Context = requireContext()
        AlertDialog.Builder(context)
                .setTitle(R.string.restore_defaults)
                .setMessage(R.string.restore_defaults_confirmation)
                .setNegativeButton(R.string.cancel, null)
                .setPositiveButton(R.string.ok, DialogInterface.OnClickListener({ dialog: DialogInterface?, which: Int ->
                    sharedPreferences!!.edit().remove(savedInstanceListKey).apply()
                    selectInstance(PeertubeInstance.DEFAULT_INSTANCE)
                    instanceListAdapter!!.submitList(PeertubeHelper.getInstanceList(context))
                }))
                .show()
    }

    private fun showAddItemDialog(c: Context) {
        val dialogBinding: DialogEditTextBinding = DialogEditTextBinding.inflate(getLayoutInflater())
        dialogBinding.dialogEditText.setInputType(
                InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_URI)
        dialogBinding.dialogEditText.setHint(R.string.peertube_instance_add_help)
        AlertDialog.Builder(c)
                .setTitle(R.string.peertube_instance_add_title)
                .setIcon(R.drawable.ic_placeholder_peertube)
                .setView(dialogBinding.getRoot())
                .setNegativeButton(R.string.cancel, null)
                .setPositiveButton(R.string.ok, DialogInterface.OnClickListener({ dialog1: DialogInterface?, which: Int ->
                    val url: String = dialogBinding.dialogEditText.getText().toString()
                    addInstance(url)
                }))
                .show()
    }

    private fun addInstance(url: String) {
        val cleanUrl: String? = cleanUrl(url)
        if (cleanUrl == null) {
            return
        }
        binding!!.loadingProgressBar.setVisibility(View.VISIBLE)
        val disposable: Disposable = Single.fromCallable(Callable({
            val instance: PeertubeInstance = PeertubeInstance(cleanUrl)
            instance.fetchInstanceMetaData()
            instance
        })).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
                .subscribe(Consumer({ instance: PeertubeInstance ->
                    binding!!.loadingProgressBar.setVisibility(View.GONE)
                    add(instance)
                }), Consumer({ e: Throwable? ->
                    binding!!.loadingProgressBar.setVisibility(View.GONE)
                    Toast.makeText(getActivity(), R.string.peertube_instance_add_fail,
                            Toast.LENGTH_SHORT).show()
                }))
        disposables!!.add(disposable)
    }

    private fun cleanUrl(url: String): String? {
        var cleanUrl: String = url.trim({ it <= ' ' })
        // if protocol not present, add https
        if (!cleanUrl.startsWith("http")) {
            cleanUrl = "https://" + cleanUrl
        }
        // remove trailing slash
        cleanUrl = cleanUrl.replace("/$".toRegex(), "")
        // only allow https
        if (!cleanUrl.startsWith("https://")) {
            Toast.makeText(getActivity(), R.string.peertube_instance_add_https_only,
                    Toast.LENGTH_SHORT).show()
            return null
        }
        // only allow if not already exists
        for (instance: PeertubeInstance? in instanceListAdapter!!.getCurrentList()) {
            if ((instance!!.getUrl() == cleanUrl)) {
                Toast.makeText(getActivity(), R.string.peertube_instance_add_exists,
                        Toast.LENGTH_SHORT).show()
                return null
            }
        }
        return cleanUrl
    }

    private fun add(instance: PeertubeInstance) {
        val list: ArrayList<PeertubeInstance?> = ArrayList(instanceListAdapter!!.getCurrentList())
        list.add(instance)
        instanceListAdapter!!.submitList(list)
    }

    private fun getItemTouchCallback(): ItemTouchHelper.SimpleCallback {
        return object : ItemTouchHelper.SimpleCallback(ItemTouchHelper.UP or ItemTouchHelper.DOWN,
                ItemTouchHelper.START or ItemTouchHelper.END) {
            public override fun interpolateOutOfBoundsScroll(recyclerView: RecyclerView,
                                                             viewSize: Int,
                                                             viewSizeOutOfBounds: Int,
                                                             totalSize: Int,
                                                             msSinceStartScroll: Long): Int {
                val standardSpeed: Int = super.interpolateOutOfBoundsScroll(recyclerView, viewSize,
                        viewSizeOutOfBounds, totalSize, msSinceStartScroll)
                val minimumAbsVelocity: Int = max(12.0, abs(standardSpeed.toDouble())).toInt()
                return minimumAbsVelocity * sign(viewSizeOutOfBounds.toDouble()).toInt()
            }

            public override fun onMove(recyclerView: RecyclerView,
                                       source: RecyclerView.ViewHolder,
                                       target: RecyclerView.ViewHolder): Boolean {
                if ((source.getItemViewType() != target.getItemViewType()
                                || instanceListAdapter == null)) {
                    return false
                }
                val sourceIndex: Int = source.getBindingAdapterPosition()
                val targetIndex: Int = target.getBindingAdapterPosition()
                instanceListAdapter!!.swapItems(sourceIndex, targetIndex)
                return true
            }

            public override fun isLongPressDragEnabled(): Boolean {
                return false
            }

            public override fun isItemViewSwipeEnabled(): Boolean {
                return true
            }

            public override fun onSwiped(viewHolder: RecyclerView.ViewHolder,
                                         swipeDir: Int) {
                val position: Int = viewHolder.getBindingAdapterPosition()
                // do not allow swiping the selected instance
                if ((instanceListAdapter!!.getCurrentList().get(position)!!.getUrl()
                                == selectedInstance!!.getUrl())) {
                    instanceListAdapter!!.notifyItemChanged(position)
                    return
                }
                val list: ArrayList<PeertubeInstance?> = ArrayList(instanceListAdapter!!.getCurrentList())
                list.removeAt(position)
                if (list.isEmpty()) {
                    list.add(selectedInstance)
                }
                instanceListAdapter!!.submitList(list)
            }
        }
    }

    /*//////////////////////////////////////////////////////////////////////////
    // List Handling
    ////////////////////////////////////////////////////////////////////////// */
    private inner class InstanceListAdapter internal constructor(context: Context?, private val itemTouchHelper: ItemTouchHelper?) : ListAdapter<PeertubeInstance?, InstanceListAdapter.TabViewHolder>(PeertubeInstanceCallback()) {
        private val inflater: LayoutInflater
        private var lastChecked: RadioButton? = null

        init {
            inflater = LayoutInflater.from(context)
        }

        fun swapItems(fromPosition: Int, toPosition: Int) {
            val list: ArrayList<PeertubeInstance?> = ArrayList(getCurrentList())
            Collections.swap(list, fromPosition, toPosition)
            submitList(list)
        }

        public override fun onCreateViewHolder(parent: ViewGroup,
                                               viewType: Int): TabViewHolder {
            return TabViewHolder(ItemInstanceBinding.inflate(inflater,
                    parent, false))
        }

        public override fun onBindViewHolder(holder: TabViewHolder,
                                             position: Int) {
            holder.bind(position)
        }

        internal inner class TabViewHolder(private val itemBinding: ItemInstanceBinding) : RecyclerView.ViewHolder(binding!!.getRoot()) {
            @SuppressLint("ClickableViewAccessibility")
            fun bind(position: Int) {
                itemBinding.handle.setOnTouchListener(OnTouchListener({ view: View?, motionEvent: MotionEvent ->
                    if (motionEvent.getActionMasked() == MotionEvent.ACTION_DOWN) {
                        if (itemTouchHelper != null && getItemCount() > 1) {
                            itemTouchHelper.startDrag(this)
                            return@setOnTouchListener true
                        }
                    }
                    false
                }))
                val instance: PeertubeInstance? = getItem(position)
                itemBinding.instanceName.setText(instance!!.getName())
                itemBinding.instanceUrl.setText(instance.getUrl())
                itemBinding.selectInstanceRB.setOnCheckedChangeListener(null)
                if ((selectedInstance!!.getUrl() == instance.getUrl())) {
                    if (lastChecked != null && lastChecked !== itemBinding.selectInstanceRB) {
                        lastChecked!!.setChecked(false)
                    }
                    itemBinding.selectInstanceRB.setChecked(true)
                    lastChecked = itemBinding.selectInstanceRB
                }
                itemBinding.selectInstanceRB.setOnCheckedChangeListener(CompoundButton.OnCheckedChangeListener({ buttonView: CompoundButton?, isChecked: Boolean ->
                    if (isChecked) {
                        selectInstance(instance)
                        if (lastChecked != null && lastChecked !== itemBinding.selectInstanceRB) {
                            lastChecked!!.setChecked(false)
                        }
                        lastChecked = itemBinding.selectInstanceRB
                    }
                }))
                itemBinding.instanceIcon.setImageResource(R.drawable.ic_placeholder_peertube)
            }
        }
    }

    private class PeertubeInstanceCallback() : DiffUtil.ItemCallback<PeertubeInstance>() {
        public override fun areItemsTheSame(oldItem: PeertubeInstance,
                                            newItem: PeertubeInstance): Boolean {
            return (oldItem.getUrl() == newItem.getUrl())
        }

        public override fun areContentsTheSame(oldItem: PeertubeInstance,
                                               newItem: PeertubeInstance): Boolean {
            return ((oldItem.getName() == newItem.getName()) && (oldItem.getUrl() == newItem.getUrl()))
        }
    }
}
